### PR TITLE
feat: Add Batch Write

### PR DESCRIPTION
**Pull request by sandercm**
Added the possibility of using batchWrite to quickly execute 25 put/delete requests on multiple different tables at once.

### CODE REVIEW

File: readme.md
Comment (GentileFulvio) [inline readme.md]:
```suggestion
			Table1.createBatchPutItem(
				{partitionKey: 'PK', sortKey: 'SK'},
				{name: 'Sander', lastname: 'Machado'}
			),
```

File: readme.md
Comment (GentileFulvio) [inline readme.md]:
```suggestion
			Table2.createBatchDeleteItem(
				{partitionKey: '123', sortKey: '456'}
			),
```

File: src/test/methods/batch-write.ts
Comment (GentileFulvio) [inline src/test/methods/batch-write.ts]: indentation nitpicking
```suggestion
			Table1.createBatchPutItem(
				{partitionKey: 'PK', sortKey: 'SK'},
				{name: 'name', lastname: 'lastname'}
			),
			Table1.createBatchPutItem(
				{partitionKey: 'PK', sortKey: 'SK23'},
				{name: 'name23', lastname: 'lastname23'}
			),
			Table2.createBatchDeleteItem(
				{partitionKey: 'PK2', sortKey: 'SK2'}
			),
			Table2.createBatchDeleteItem(
				{partitionKey: 'PK2', sortKey: 'SK3'}
			),
			Table2.createBatchPutItem(
				{partitionKey: 'PK', sortKey: 'SK'},
				{name: 'name', lastname: 'lastname'}
			)
		]
```

File: src/test/methods/batch-write.ts
Comment (GentileFulvio) [inline src/test/methods/batch-write.ts]: How does this test actually test that it has beem retried? We could check `updateStub.callCount` before and after we call the batchWrite for example

**Review by GentileFulvio — CHANGES_REQUESTED** [review]
Had a quick look already. Still need to check thoroughly

File: src/test/methods/batch-write.ts
Comment (sandercm) [reply src/test/methods/batch-write.ts]: I agree, I will extend this test to properly test the exponential backoff.

Comment (sandercm) [reply src/test/methods/batch-write.ts]: The test will now check that every functional call has happened and was called the correct arguments.

File: src/lib/methods/insert-item.ts
Comment (GentileFulvio) [inline src/lib/methods/insert-item.ts]: Revert all the changes

File: src/lib/methods/batch-put.ts
Comment (GentileFulvio) [inline src/lib/methods/batch-put.ts]: Add newline below

File: src/lib/table.ts
Comment (GentileFulvio) [inline src/lib/table.ts:103]: newline below

File: src/lib/methods/batch-write.ts
Comment (GentileFulvio) [inline src/lib/methods/batch-write.ts]: Move this import to the top

File: src/lib/methods/base-method.ts
Comment (GentileFulvio) [inline src/lib/methods/base-method.ts]: Move import up

**Review by GentileFulvio — CHANGES_REQUESTED** [review]

File: readme.md
Comment (bartcallant) [inline readme.md]: To which normal action does this `put` refer? Is it an `insert`, `upsert` or `update`? Or is that non mappable to a specific action?

How does the `createBatchDeleteItem` defer from a "normal" `DeleteItem`?

The reason I'm asking is that these actions seem quite a bit different from the existing actions. Which might make it difficult for a user to know the difference between the two actions.  The `transactWrite` action for example, does something similar to the `batchWrite`, but it accepts the "normal" actions.
Would it be possible that the `batchWrite` also accepts the "normal" actions? I think it will make the public API more consistent. Internally it could still use those other actions to separate the the logic.

File: src/lib/methods/method.ts
Comment (bartcallant) [inline src/lib/methods/method.ts:6]: Is there any benefit of this? Since this does not have custom logic compared to the `BaseMethod`?

File: src/lib/methods/batch-write.ts
Comment (bartcallant) [inline src/lib/methods/batch-write.ts]: Maybe put all the files related to `BatchXX` in a separate folder. Kinda like `transactions`.

**Review by bartcallant — CHANGES_REQUESTED** [review]
Did a first review. Very nice this is finally being added :ok_hand:

File: src/lib/methods/method.ts
Comment (sandercm) [reply src/lib/methods/method.ts:6]: This is mostly for Typings, as Method has a field of type Params while BatchMethod has a field of type BatchParams. This seemed like the cleanest solution without having to do major refactoring or breaking changes. Like this we keep all the shared logic together while we can separate the differing interface.

File: readme.md
Comment (sandercm) [reply readme.md]: This maps to the `put` action which currently isn't used in the project. The big difference between `batch` and `normal` operations is that batch operations don't take conditionals. So you can't use `$where` etc. Currently the library doesn't actually use the `put` action build into DynamoDB that's why this seems so different. It matches the `insert` way of doing things `insert({key}, {obj})`. Having to convert it from `{key, obj}` -> `update expression` -> `{key, obj}` for every item in the batch seems like a big waste? You would also lose the strong typing that batchInsert has now as `method` uses `any` for it's interface.

File: src/lib/methods/batch-write.ts
Comment (sandercm) [reply src/lib/methods/batch-write.ts]: good point

File: src/lib/dynamodb.ts
Comment (SamVerschueren) [inline src/lib/dynamodb.ts]: I think it would be better if we had the same syntax as transactions. So instead of passing an array, it's just an argument list.

```suggestion
	batchWrite(...items: BatchItem[]) {
```

Then the example will look like this so make sure to update the documentation.

```ts
const result = await db.batchWrite(
    Table1.createBatchPutItem(
        {partitionKey: 'PK', sortKey: 'SK'},
        {name: 'Sander', lastname: 'Machado'}
    ),
    Table1.createBatchPutItem(
        {partitionKey: 'PK', sortKey: 'SK23'},
        {name: 'Sander', lastname: 'Doe'}
    )
);
```

File: src/test/methods/batch-write.ts
Comment (SamVerschueren) [inline src/test/methods/batch-write.ts]: Just some extra newlines :).

```suggestion
});

test.serial('program should use exponential backoff', async t => {
	const beforeCount = updateStub.callCount;

	await db.batchWrite(
		[
			Table1.createBatchPutItem(
				{partitionKey: 'PK', sortKey: 'SK'},
				{name: 'name', lastname: 'lastname'}
			)
		]
	).exec();

	const afterCount = updateStub.callCount;

	t.deepEqual(afterCount - beforeCount, 4);

	const request = {
		RequestItems: {
			'insert-Table1': [
				{
					PutRequest: {
						Item: {
							partitionKey: 'PK',
							sortKey: 'SK',
							name: 'name',
							lastname: 'lastname'
						}
					}
				}
			]
		}
	};

	for (let i = 4; i > 0; i--) {
		t.deepEqual(updateStub.getCall(i).lastArg, request);
	}

	t.deepEqual(updateStub.lastCall.args[0], request);
});
```

File: src/test/methods/batch-write.ts
Comment (SamVerschueren) [inline src/test/methods/batch-write.ts]: The `ts-ignore` should be necessary. Just type the array correctly

```ts
const batchItems: PutRequest[] = [];
```

File: src/test/methods/batch-write.ts
Comment (SamVerschueren) [inline src/test/methods/batch-write.ts]:
```suggestion

	await t.throwsAsync(() => db.batchWrite(batchItems).exec(), 'Can not insert more than 25 items at a time.');
```

File: src/test/methods/batch-write.ts
Comment (SamVerschueren) [inline src/test/methods/batch-write.ts]:
```suggestion
	ItemCollectionMetrics: null,
	ConsumedCapacity: null,
	UnprocessedItems: {'insert-Table1': [
			{
				PutRequest: {
					Item: {
						partitionKey: 'PK',
						sortKey: 'SK',
						name: 'name',
						lastname: 'lastname'
					}
				}
			}
		]
	}
};
```

File: src/lib/table.ts
Comment (SamVerschueren) [inline src/lib/table.ts]:
```suggestion
	 * @param  key			The primary key of the record we want to insert.
```

File: src/lib/methods/batch/batch-write.ts
Comment (SamVerschueren) [inline src/lib/methods/batch/batch-write.ts]: Use a regular `for-of`. I don't like `forEach` 🤷 :).

```suggestion
		this.items.forEach((item) => {
			const table = request.RequestItems[item.table];
			if (!table) {
				request.RequestItems[item.table] = [item.value];
			} else {
				(request.RequestItems[item.table] as any[]).push(item.value);
			}
		});
```

File: src/lib/methods/batch/batch-write.ts
Comment (SamVerschueren) [inline src/lib/methods/batch/batch-write.ts]:
```suggestion
	}

	/**
```

File: src/lib/methods/batch/batch-write.ts
Comment (SamVerschueren) [inline src/lib/methods/batch/batch-write.ts]:
```suggestion
	}

	/**
	 * Execute the batch write request.
```

File: src/lib/methods/batch/batch-write.ts
Comment (SamVerschueren) [inline src/lib/methods/batch/batch-write.ts]: Any reason we can't just use `await delay(this.msDelay)`?

File: src/lib/methods/batch/batch-write.ts
Comment (SamVerschueren) [inline src/lib/methods/batch/batch-write.ts]: I'm wondering if this entire block can't be rewritten with [p-retry](https://github.com/sindresorhus/p-retry#readme) which has built-in exponential back-off and works properly with the provided retry options.

**Review by SamVerschueren — CHANGES_REQUESTED** [review]
Added some feedback. 

Very nice to see this being added 🙏 !

File: src/lib/methods/batch/batch-write.ts
Comment (sandercm) [reply src/lib/methods/batch/batch-write.ts]: With @GentileFulvio help we managed to refactor this to use pRetry.

File: src/lib/methods/batch/batch-write.ts
Comment (GentileFulvio) [inline src/lib/methods/batch/batch-write.ts]: Perhaps we should call this `randomizeRetry` or something if we really want to provide this to the constructor. It isn't really clear what this `randomize` actually randomizes imho

**Review by GentileFulvio — CHANGES_REQUESTED** [review]
One minor comment. Besides that I think this one is good to go if we can get the approval of @SamVerschueren

File: readme.md
Comment (SamVerschueren) [inline readme.md]:
```suggestion
const result = await db.batchWrite(
	Table1.createBatchPutItem(
		{partitionKey: 'PK', sortKey: 'SK'},
		{name: 'Sander', lastname: 'Machado'}
	),
	Table1.createBatchPutItem(
		{partitionKey: 'PK', sortKey: 'SK23'},
		{name: 'Sander', lastname: 'Doe'}
	),
	Table2.createBatchDeleteItem(
		{partitionKey: '123', sortKey: '456'}
	),
	Table2.createBatchDeleteItem(
		{partitionKey: 'PK2', sortKey: 'SK3'}
	),
	Table2.createBatchPutItem(
		{partitionKey: 'PK', sortKey: 'SK'},
		{name: 'name', lastname: 'lastname'}
	)
).exec();
```

File: src/lib/methods/batch/batch-write.ts
Comment (SamVerschueren) [inline src/lib/methods/batch/batch-write.ts]:
```suggestion
				const {UnprocessedItems} = await db.batchWrite(query).promise();

				if (UnprocessedItems && Object.keys(UnprocessedItems).length > 0) {
```

File: src/lib/methods/batch/batch-write.ts
Comment (SamVerschueren) [inline src/lib/methods/batch/batch-write.ts]: Maybe we should add a message here indicating that we still have `unprocessedItems.length` items to process.

```suggestion
					throw new Error(`${Object.keys(UnprocessedItems).length} could not be processed`);
```

File: src/test/methods/batch-write.ts
Comment (SamVerschueren) [inline src/test/methods/batch-write.ts]:
```suggestion
	await db.batchWrite(
		Table1.createBatchPutItem(
			{partitionKey: 'PK', sortKey: 'SK'},
			{name: 'name', lastname: 'lastname'}
		),
		Table1.createBatchPutItem(
			{partitionKey: 'PK', sortKey: 'SK23'},
			{name: 'name23', lastname: 'lastname23'}
		),
		Table2.createBatchDeleteItem(
			{partitionKey: 'PK2', sortKey: 'SK2'}
		),
		Table2.createBatchDeleteItem(
			{partitionKey: 'PK2', sortKey: 'SK3'}
		),
		Table2.createBatchPutItem(
			{partitionKey: 'PK', sortKey: 'SK'},
			{name: 'name', lastname: 'lastname'}
		)
	).exec();
```

File: src/test/methods/batch-write.ts
Comment (SamVerschueren) [inline src/test/methods/batch-write.ts]:
```suggestion
	t.deepEqual(updateStub.lastCall.args[0], {
		RequestItems: {
			'insert-Table1': [
				{
					PutRequest: {
						Item: {
							partitionKey: 'PK',
							sortKey: 'SK',
							name: 'name',
							lastname: 'lastname'
						}
					}
				},
				{
					PutRequest: {
						Item: {
							partitionKey: 'PK',
							sortKey: 'SK23',
							name: 'name23',
							lastname: 'lastname23'
						}
					}
				}
			],
			'insert-Table2': [
				{
					DeleteRequest: {
						Key: {partitionKey: 'PK2', sortKey: 'SK2'}
					}
				},
				{
					DeleteRequest: {
						Key: {partitionKey: 'PK2', sortKey: 'SK3'}
					}
				},
				{
					PutRequest: {
						Item: {
							partitionKey: 'PK',
							sortKey: 'SK',
							name: 'name',
							lastname: 'lastname'
						}
					}
				}
			]
		}
	});
```

File: src/test/methods/batch-write.ts
Comment (SamVerschueren) [inline src/test/methods/batch-write.ts]:
```suggestion
	await db.batchWrite(
		Table1.createBatchPutItem(
			{partitionKey: 'PK', sortKey: 'SK'},
			{name: 'name', lastname: 'lastname'}
		)
	).exec();
```

**Review by SamVerschueren — CHANGES_REQUESTED** [review]
Firs round of review with some suggestions (I'll quickly apply them myself).

File: src/test/methods/batch-write.ts
Comment (SamVerschueren) [inline src/test/methods/batch-write.ts]:
```suggestion
	UnprocessedItems: {
		'insert-Table1': [
			{
				PutRequest: {
					Item: {
						partitionKey: 'PK',
						sortKey: 'SK',
						name: 'name',
						lastname: 'lastname'
					}
				}
			}
		]
	}
```

File: src/test/methods/batch-write.ts
Comment (SamVerschueren) [inline src/test/methods/batch-write.ts]:
```suggestion
};

const throttleFixture = {
```

File: src/test/methods/batch-write.ts
Comment (SamVerschueren) [inline src/test/methods/batch-write.ts]:
```suggestion
};

const response = {
```

File: src/test/methods/batch-write.ts
Comment (SamVerschueren) [inline src/test/methods/batch-write.ts]:
```suggestion
};

// @ts-ignore
```

File: src/test/methods/batch-write.ts
Comment (SamVerschueren) [inline src/test/methods/batch-write.ts]:
```suggestion
};

const sandbox = sinon.createSandbox();
```

**Review by SamVerschueren — CHANGES_REQUESTED** [review]

File: src/lib/methods/batch/batch-write.ts
Comment (SamVerschueren) [inline src/lib/methods/batch/batch-write.ts]:
```suggestion
	}
}
```

File: src/lib/methods/batch/batch-write.ts
Comment (SamVerschueren) [inline src/lib/methods/batch/batch-write.ts]:
```suggestion
	}

	/**
```

File: src/lib/methods/batch/batch-write.ts
Comment (SamVerschueren) [inline src/lib/methods/batch/batch-write.ts]: Just wondering. Why do we have the `initialize` here? Can't we move the code from the `initialize` to the `buildRawQuery`?

Also, is there any reason the request is assigned to `this.params`? It feels like `this.params` might not even have to exist?

File: src/lib/methods/batch/batch-item.ts
Comment (SamVerschueren) [inline src/lib/methods/batch/batch-item.ts]: I would make this `abstract buildRawQuery(): BatchItemQuery` where `BatchItemQuery` is of type

```ts
interface BatchItemQuery {
	PutRequest: {
		Item: {
			[key: string]: any;
		};
	};
	DeleteRequest: {
		Item: {
			[key: string]: any;
		};
	};
}
```

This is more in line with how all the other query stuff works. All methods just have a `buildRawQuery()` which returns the query from that item.

File: src/lib/methods/batch/batch-put.ts
Comment (SamVerschueren) [inline src/lib/methods/batch/batch-put.ts]: Rename to `buildRawQuery()` as mentioned above.

File: src/lib/methods/batch/batch-write.ts
Comment (SamVerschueren) [inline src/lib/methods/batch/batch-write.ts]: I would drop `this.params` and just return the `request` directly.

File: src/lib/methods/batch/batch-delete.ts
Comment (SamVerschueren) [inline src/lib/methods/batch/batch-delete.ts]: See comment below, I would rename this to `buildRawQuery`.

**Review by SamVerschueren — CHANGES_REQUESTED** [review]

File: src/lib/methods/batch/batch-write.ts
Comment (SamVerschueren) [inline src/lib/methods/batch/batch-write.ts]: I believe this block can be rewritten with the `runQuery` method which does the retry handling entirely for you.

```suggestion
		return this.runQuery(() => {
			const {UnprocessedItems} = await db.batchWrite(query).promise();

		    if (UnprocessedItems && Object.keys(UnprocessedItems).length > 0) {
		    	query = {RequestItems: UnprocessedItems};

		    	throw new Error(`${Object.keys(UnprocessedItems).length} could not be processed`);
		    }
		});
```

The retry configuration should be set on the `BaseMethod` with the `.retry()` which has defaults as can be seen in `utils/retry-options.ts`.

File: src/lib/methods/batch/batch-write.ts
Comment (SamVerschueren) [inline src/lib/methods/batch/batch-write.ts]:
```suggestion
	constructor(
		dynamodb: DynamoDB,
		private items: BatchItem[],
	) {
		super(dynamodb);
	}
```

As explained below, the retry configuration should be set with the `.retry()` method.

**Review by SamVerschueren — CHANGES_REQUESTED** [review]

File: src/lib/methods/batch/batch-write.ts
Comment (sandercm) [reply src/lib/methods/batch/batch-write.ts]: Yes I agree we should remove initialize and replace it with buildRawQuery. I think I was confusing the 2 when looking at the other classes. The this.params approach was used because I wanted it to be similar to the other classes. I replaced it now with the suggested changes and made buildRawQuery return BatchWriteItemInput which keeps the typechecking.

File: src/lib/methods/batch/batch-write.ts
Comment (sandercm) [reply src/lib/methods/batch/batch-write.ts]: I updated the class to use the retry() syntax but for some reason, I can not get this.runQuery() to actually retry the function calls only using pRetry.

Comment (GentileFulvio) [reply src/lib/methods/batch/batch-write.ts]: The issue is that the call doesn't fail @SamVerschueren it always resolves but returns unprocessedItems. Are we sure it's safe to do this with the `this.runQuery` Don't we risk that the logic will keep retrying the same items over and over instead of just retrying the `unprocessed` ones ?

File: src/lib/errors/UnprocessedItems.ts
Comment (SamVerschueren) [inline src/lib/errors/UnprocessedItems.ts]: This is a bit different than we do it in the monorepo.

```ts
export class UnprocessedItemsException extends Error {
	readonly name: 'UnprocessedItemsException' = 'UnprocessedItemsException';
	readonly code: 'UnprocessedItemsException' = 'UnprocessedItemsException';

	constructor(message: string) {
		super(message);

		// Restore prototype chain
		Object.setPrototypeOf(this, new.target.prototype);
	}
}
```

this should allow you to say that `error instanceof UnprocessedItemsException` which is not possible with the current implementation.

Comment (SamVerschueren) [reply src/lib/errors/UnprocessedItems.ts]: Also, rename the file to `unprocessed-items-exception.ts` ;).

**Review by SamVerschueren — CHANGES_REQUESTED** [review]
Only the custom error. Other than that, this looks 👍 !!

**Review by SamVerschueren — APPROVED** [review]
cheers

**Comment by onhate** [comment]
NICE! I was looking for this

**Pull request merged**
